src/System.Linq.Expressions/src/System/Dynamic/UpdateDelegates.Generated.tt

```diff
@@ -0,0 +1,332 @@
+
+                //
+                // Execute the rule on the matchmaker site
+                //
+
+                try
+                {
+                    result = rule(<#=ruleInvocationArguments#>);
+                    if (CallSiteOps.GetMatch(site))
+                    {
+                        return result;
+                    }
+                }
+                finally
+                {
+                    if (CallSiteOps.GetMatch(site))
+                    {
+                        //
+                        // The rule worked. Add it to level 1 cache.
+                        //
+                        CallSiteOps.AddRule(@this, rule);
+                    }
+                }
+
+                // Rule we got back didn't work, try another one
+                CallSiteOps.ClearMatch(site);
+            }
+        }
+
+        [Obsolete("pregenerated CallSite<T>.Update delegate", error: true)]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Usage", "CA1801:ReviewUnusedParameters")]
+        internal static TRet NoMatch<#=i#><<#=genericParameters#>>(<#=parameters#>)
+        {
+            site._match = false;
+            return default(TRet);
+        }
+
+
+
+<#
+}
+#>
+<#
+for (int i = 1; i <= 10; i++)
+{
+    string genericParameters = string.Join(", ", Enumerable.Range(0, i).Select(j => "T" + j));
+    string parameters = string.Join(", ", new[] { "CallSite site" }.Concat(Enumerable.Range(0, i).Select(j => "T" + j + " arg" + j)));
+    string ruleInvocationArguments = string.Join(", ", new[] { "site" }.Concat(Enumerable.Range(0, i).Select(j => "arg" + j)));
+    string argsArray = i == 0 ? "Array.Empty<object>()" : "new object[] { " + string.Join(", ", Enumerable.Range(0, i).Select(j => "arg" + j)) + " }";
+#>
+        [Obsolete("pregenerated CallSite<T>.Update delegate", error: true)]
+        internal static void UpdateAndExecuteVoid<#=i#><<#=genericParameters#>>(<#=parameters#>)
+        {
+            //
+            // Declare the locals here upfront. It actually saves JIT stack space.
+            //
+            var @this = (CallSite<Action<CallSite, <#=genericParameters#>>>)site;
+            Action<CallSite, <#=genericParameters#>>[] applicable;
+            Action<CallSite, <#=genericParameters#>> rule, originalRule = @this.Target;
+
+
+            //
+            // Create matchmaker and its site. We'll need them regardless.
+            //
+            site = CallSiteOps.CreateMatchmaker(@this);
+
+            //
+            // Level 1 cache lookup
+            //
+            if ((applicable = CallSiteOps.GetRules(@this)) != null)
+            {
+                for (int i = 0; i < applicable.Length; i++)
+                {
+                    rule = applicable[i];
+
+                    //
+                    // Execute the rule
+                    //
+
+                    // if we've already tried it skip it...
+                    if ((object)rule != (object)originalRule)
```

Comment:
This is a typical pattern to ensure a reference compare - in case if arguments define equality, which System.Delegate does. 
Reference compare seems intentional here.